File: apps/api/liderix_api/schemas/notifications.py
from __future__ import annotations

from datetime import datetime
from typing import Optional, List
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from liderix_api.enums import NotificationType, NotificationStatus


class NotificationRead(BaseModel):
    id: UUID
    org_id: Optional[UUID] = None
    user_id: UUID
    type: NotificationType
    status: NotificationStatus
    title: str
    message: str
    related_entity_type: Optional[str] = None
    related_entity_id: Optional[UUID] = None
    action_url: Optional[str] = None
    action_text: Optional[str] = None
    priority: Optional[str] = None
    created_at: datetime
    read_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True, extra="forbid")


class NotificationListResponse(BaseModel):
    items: List[NotificationRead]
    page: int
    page_size: int
    total: int
    unread_count: int = 0

    model_config = ConfigDict(extra="forbid")


class NotificationCreate(BaseModel):
    org_id: Optional[UUID] = None
    user_id: UUID
    type: NotificationType
    title: str = Field(..., max_length=500)
    message: str
    related_entity_type: Optional[str] = None
    related_entity_id: Optional[UUID] = None
    action_url: Optional[str] = Field(default=None, max_length=1000)
    action_text: Optional[str] = Field(default=None, max_length=100)
    priority: Optional[str] = Field(default="normal", max_length=20)

    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)

File: n8n/src/lib/api/notifications.ts
import { api } from "@/lib/api/config"

export type NotificationStatus = "unread" | "read" | "archived"
export type NotificationType =
  | "system"
  | "task_assigned"
  | "task_completed"
  | "task_overdue"
  | "project_update"
  | "mention"
  | "comment"
  | "deadline_reminder"
  | "okr_update"
  | "kpi_alert"
  | "invitation"

export interface NotificationItem {
  id: string
  org_id?: string | null
  user_id: string
  type: NotificationType
  status: NotificationStatus
  title: string
  message: string
  related_entity_type?: string | null
  related_entity_id?: string | null
  action_url?: string | null
  action_text?: string | null
  priority?: string | null
  created_at: string
  read_at?: string | null
}

export interface NotificationListResponse {
  items: NotificationItem[]
  page: number
  page_size: number
  total: number
  unread_count: number
}

export async function getNotifications(params: {
  orgId?: string
  page?: number
  pageSize?: number
  status?: NotificationStatus
  type?: NotificationType
}): Promise<NotificationListResponse> {
  const { orgId, page = 1, pageSize = 20, status, type } = params
  const query = new URLSearchParams()
  query.set("page", String(page))
  query.set("page_size", String(pageSize))
  if (orgId) query.set("org_id", orgId)
  if (status) query.set("status", status)
  if (type) query.set("type", type)

  const { data } = await api.get(`/notifications?${query.toString()}`)
  return data
}

export async function markNotificationRead(notificationId: string): Promise<NotificationItem> {
  const { data } = await api.patch(`/notifications/${notificationId}/read`)
  return data
}

export async function markAllNotificationsRead(orgId?: string): Promise<{ updated: number }> {
  const query = orgId ? `?org_id=${encodeURIComponent(orgId)}` : ""
  const { data } = await api.patch(`/notifications/read-all${query}`)
  return data
}

export async function deleteNotification(notificationId: string): Promise<void> {
  await api.delete(`/notifications/${notificationId}`)
}
